### PR TITLE
Client api docs

### PR DESCRIPTION
**Pull request by jeromekelleher**
This is an initial pass at the client API docs. There's a lot to be filled out, but this is the basic idea.

The build docs are available at http://ga4gh-server.readthedocs.org/en/client-api-docs/api.html

(This builds on #643).

Thoughts? @diekhans, this should be up your street --- what do you think?

### CODE REVIEW

**Comment by jeromekelleher** [comment]
This is still a work in progress (some tests need fixing and the docs need polishing) but it would be good to get some feedback on the API docs. This should be the latest version of the docs

http://ga4gh-server.readthedocs.org/en/client-api-docs/api.html

Is this something that you could use to get some real work done?

**Comment by dcolligan** [comment]
#643 is merged... rebase for review, please?

**Comment by jeromekelleher** [comment]
I've rebased this @dcolligan, although it still depends on #654 and it's not ready for full review yet. I'll try to polish it up over the next couple of days.

Any high-level feedback on how it's looking?

**Comment by dcolligan** [comment]
@jeromekelleher I merged #654 ... please rebase

**Comment by jeromekelleher** [comment]
@dcolligan --- done. The client API code needs some polishing, so don't take too much notice of the current set up.

File: ga4gh/client.py [hunk 1]
```diff
@@ -325,14 +418,125 @@ def searchReads(
...
+        url = posixpath.join(self._urlPrefix, urlSuffix)
+        params = self._getHttpParameters()
+        params.update(request.toJsonDict())
+        response = requests.get(url, params=params, verify=False)
+        self._checkResponseStatus(response)
+        return self._deserializeResponse(
+            response.text, protocol.ListReferenceBasesResponse)
+
+
+class LocalClient(Client):
```
Comment (dcolligan) [inline ga4gh/client.py]: docstring for this?

File: ga4gh/client.py [hunk 2]
```diff
@@ -14,17 +14,12 @@
 import ga4gh.exceptions as exceptions
 
 
-class HttpClient(object):
-    """
-    GA4GH Http Client
-    """
+class Client(object):
```
Comment (dcolligan) [inline ga4gh/client.py]: doc for this?

**Comment by dcolligan** [comment]
Generally looks good.  I'm not in love with the doc organization... usually there's a separate section for prose and another for class / method docs.  Here they're intertwined.

**Comment by jeromekelleher** [comment]
I've made an update here @dcolligan. Sorry, this has ended up being a bit of a mammoth patch, but stuff just kept turning up on the client side that needed fixing. It should all be done now, modulo finishing off the documentation. The client side tests are a lot stronger now, and hopefully there's a framework in place to make the better as we go on. I still need to document the classes, but I'll fill them in ASAP now.

I don't quite follow what you mean about the method documentation and the prose. Can you elaborate a bit on that?

**Comment by dcolligan** [comment]
For instance, in java, javadocs are usually in a navigable web hierarchy that is separate from the installation/tutorial/whatever pages.

**Comment by jeromekelleher** [comment]
OK, that's it. The client API documentation is very much a draft and something we'll want to work on quite a bit before it's something a normal person can work with productively. I see your point about mixing sections --- I think it's something people do a fair bit using sphinx, but we can always change it. As I say, this is very much a draft.

Apologies again for the massive patch, I'll try to keep them smaller from here.

File: ga4gh/cli.py [hunk 3]
```diff
@@ -295,7 +305,31 @@ def run(self):
             self._run(self._variantSetId)
 
 
-class SearchVariantsRunner(AbstractSearchRunner):
+class VariantFormatterMixin(object):
```
Comment (dcolligan) [inline ga4gh/cli.py]: good idea

File: ga4gh/client.py [hunk 4]
```diff
@@ -14,230 +13,218 @@
...
         patterns of the other search and get requests, and is implemented
         differently.
         """
-        url = "references/{id}/bases"
-        fullUrl = posixpath.join(self._urlPrefix, url).format(id=id_)
         request = protocol.ListReferenceBasesRequest()
         request.start = start
         request.end = end
         notDone = True
+        # We should probably use a StringIO here to make string buffering
```
Comment (dcolligan) [inline ga4gh/client.py]: add a TODO tag here

File: ga4gh/client.py [hunk 5]
```diff
@@ -325,14 +362,138 @@ def searchReads(
...
+
+    def _runSearchPageRequest(
+            self, protocolRequest, objectName, protocolResponseClass):
+        searchMethod = self._searchMethodMap[objectName]
+        responseJson = searchMethod(protocolRequest.toJsonString())
+        return self._deserializeResponse(responseJson, protocolResponseClass)
+
+    def _runListReferenceBasesPageRequest(self, id_, request):
+        requestArgs = request.toJsonDict()
+        # We need to remove end from this dict if it's not specified because
```
Comment (dcolligan) [inline ga4gh/client.py]: add a TODO tag

File: tests/end_to_end/client.py [hunk 6]
```diff
@@ -21,7 +21,7 @@ def __init__(self, serverUrl, flags=None):
         self.outFile = None
         self.errFile = None
         if flags is None:
-            self.flags = "-v"
+            self.flags = "-vv"
```
Comment (dcolligan) [inline tests/end_to_end/client.py]: why this change?

**Comment by dcolligan** [comment]
+1 aside from nits

File: tests/end_to_end/client.py (hunk 6 above)
Comment (jeromekelleher) [inline tests/end_to_end/client.py]: For some reason, the logs emitting POST messages are only there at a higher log level now and tests were failing because of this. I don't really know why this has changed, but I hope we'll be changing this code around soon enough so I didn't worry too much about it.

**Comment by jeromekelleher** [comment]
TODOs added @dcolligan.
